---
author: "Peter Diakumis UMCCR"
date: "`r format(Sys.time(), '%a %Y-%b-%d')`"
output:
  html_document:
    theme: simplex
    toc: true
    toc_float: false
    toc_depth: 5
    code_download: true
    code_folding: hide
  rmdformats::material:
    highlight: kate
params:
  woof_final: "/path/to/woof/final"
title: "`r paste('Comparison of woof runs in', htmltools::br(), params$woof_final)`"
description: "Comparison of bcbio and umccrise pipeline runs"
---

<style type="text/css">
.main-container {
  max-width: 1400px;
  margin-left: auto;
  margin-right: auto;
}
</style>


```{r load-knitr, include=F}
require(knitr)
knitr::opts_chunk$set(
  echo = TRUE
  )
```

```{r load-pkgs, message=F, warning=F}
require(bedr)
require(DiagrammeR)
require(dplyr)
require(DT)
require(glue)
require(kableExtra)
require(purrr)
require(readr)
require(rock)
require(tidyr)
require(woofr)
```

Results
-------

### SNVs/Indels

```{r funcs-snvs}
f1 <- params$f1
f2 <- params$f2
# f1 <- "bcbio_116a0_GRCh37_cromwell"
# f2 <- "bcbio_116a0_GRCh37_native"
comp_dir <- paste0(f1, "-vs-", f2)
raijin_share <- glue::glue("/g/data3/gx8/projects/Diakumis/woof_compare/{comp_dir}")
# raijin_share <- glue::glue("~/my_apps/sshfs/raijin/{comp_dir}") # locally
stopifnot(dir.exists(raijin_share))

samples <- c("2016.249.17.MH.P033",
             "2016.249.18.SV.P007",
             "2016.249.18.WH.P025",
             "CUP-Pairs8",
             "SFRC01073")

vcf_types <- c("ensemble-batch", "ensemble-germ", "gatk-germ", "mutect-batch",
               "strelka-batch", "strelka-germ", "vardict-batch", "vardict-germ",
               "pcgr", "cpsr")

read_count_files <- function(sample, vcf_type, pass_or_all) {

  get_count_file <- function(sample, f1_or_f2, vcf_type, pass_or_all) {
    stopifnot(sample %in% samples, vcf_type %in% vcf_types, f1_or_f2 %in% c("f1", "f2"), pass_or_all %in% c("PASS", "ALL"))
    x <- file.path(raijin_share, sample, "woof/final/vcf_counts", f1_or_f2, vcf_type, pass_or_all, "count_vars.txt")
    stopifnot(file.exists(x))
    x
  }

  c1 <- get_count_file(sample, "f1", vcf_type, pass_or_all) %>% readr::read_lines() %>% as.integer()
  c2 <- get_count_file(sample, "f2", vcf_type, pass_or_all) %>% readr::read_lines() %>% as.integer()
  dplyr::tibble(run1_count = c1, run2_count = c2)
}

read_eval_file <- function(sample, vcf_type, pass_or_all) {
  stopifnot(sample %in% samples, vcf_type %in% vcf_types, pass_or_all %in% c("PASS", "ALL"))
  file.path(raijin_share, sample, "woof/final/vcf_eval", vcf_type, pass_or_all, "eval_stats.tsv") %>%
    readr::read_tsv(col_types = cols(.default = "d"))
}

get_stats <- function(sample, vcf_type, pass_or_all) {
  stopifnot(sample %in% samples, vcf_type %in% vcf_types, pass_or_all %in% c("PASS", "ALL"))
  cbind(
    read_count_files(sample, vcf_type, pass_or_all),
    read_eval_file(sample, vcf_type, pass_or_all))

}

```

```{r analysis-snvs}
get_stats_sample <- function(sample, pass_or_all) {
  vcf_types %>%
    purrr::map(~ get_stats(sample, .x, pass_or_all)) %>%
    purrr::set_names(vcf_types) %>%
    dplyr::bind_rows(.id = "vcf_type") %>%
    dplyr::mutate(subset = pass_or_all) %>%
    dplyr::select(subset, everything())
}


d_pass <- samples %>%
  purrr::map(~ get_stats_sample(., "PASS")) %>%
  purrr::set_names(samples) %>%
  dplyr::bind_rows(.id = "sample")

d_all <- samples %>%
  purrr::map(~ get_stats_sample(., "ALL")) %>%
  purrr::set_names(samples) %>%
  dplyr::bind_rows(.id = "sample")

d <- dplyr::bind_rows(d_all, d_pass)

```


```{r tab-descr-snvs}

x <- dplyr::tribble(
  ~Column,    ~Description,
  "run1",     f1,
  "run2",     paste(f2, "(truthset)"),
  "subset",   "All/PASSed variants",
  "SNP",      "SNP results",
  "IND",      "Indel results",
  "FP",       "False Positive",
  "FN",       "False Negative",
  "TP",       "True Positive",
  "Truth",    "TP + FN",
  "Recall",   "TP / Truth",
  "Precision", "TP / TP + FP"
)

cap <- x %>%
  mutate(d = paste(Column, Description, sep = ": ")) %>%
  pull(d)
cap <- htmltools::tags$caption(htmltools::div(paste(cap[1:5], collapse = "; ")),
                               htmltools::div(paste(cap[6:length(cap)], collapse = "; ")))
```

#### Comparison Table

```{r eval-tab-snvs}
bgsize <- '90% 90%'
big_mark_cols <- c("run1_count", "run2_count",
                   "SNP_Truth", "SNP_TP", "SNP_FP", "SNP_FN",
                   "IND_Truth", "IND_TP", "IND_FP", "IND_FN")
d %>%
  dplyr::select(-matches("f[123]$")) %>%
  dplyr::select(sample, vcf_type, subset, run1_count, run2_count,
                contains("Recall"), contains("Precision"),
                everything()) %>%
  dplyr::mutate_if(is.numeric, round, 3) %>%
  dplyr::mutate_if(is.character, as.factor) %>%
  DT::datatable(
    rownames = FALSE, caption = cap, filter = list(position = "top", clear = FALSE, plain = TRUE),
    extensions = c("Scroller"), options = list(scroller = TRUE, scrollX = TRUE, scrollY = 600)) %>%
  formatStyle(
    'SNP_Recall',
    background = styleColorBar(d$SNP_Recall, 'lightgreen'),
    backgroundSize = bgsize, backgroundRepeat = 'no-repeat', backgroundPosition = 'center'
  ) %>%
  formatStyle(
    'IND_Recall',
    background = styleColorBar(d$IND_Recall, 'lightgreen'),
    backgroundSize = bgsize, backgroundRepeat = 'no-repeat', backgroundPosition = 'center'
  ) %>%
  formatStyle(
    'SNP_Precision',
    background = styleColorBar(d$SNP_Precision, 'lightgreen'),
    backgroundSize = bgsize, backgroundRepeat = 'no-repeat', backgroundPosition = 'center'
  ) %>%
  formatStyle(
    'IND_Precision',
    background = styleColorBar(d$IND_Precision, 'lightgreen'),
    backgroundSize = bgsize, backgroundRepeat = 'no-repeat', backgroundPosition = 'center'
  ) %>%
  formatCurrency(big_mark_cols, currency = "", interval = 3, mark = ",", digits = 0)
```

#### PCGR FP/FN Tiers

```{r pcgr-read-fpfn}
read_pcgr_fpfn <- function(sample) {

  read_pcgr_fp_or_fn <- function(sample, fp_or_fn) {
    stopifnot(sample %in% samples, length(sample) == 1,
              fp_or_fn %in% c("fp", "fn"), length(fp_or_fn) == 1)

    fp_or_fn <- ifelse(fp_or_fn == "fp", "0000.vcf", "0001.vcf")
    x <- file.path(raijin_share, sample, "woof/final/bcftools_isec/pcgr/ALL", fp_or_fn)
    bedr::read.vcf(x, split.info = TRUE, verbose = FALSE)$vcf %>%
      tibble::as_tibble() %>%
      dplyr::select(CHROM, POS, PCGR_TIER, PCGR_SYMBOL, QUAL, AF, CALLERS)
  }

  fp <- read_pcgr_fp_or_fn(sample, "fp") %>% dplyr::mutate(fp_or_fn = "fp")
  fn <- read_pcgr_fp_or_fn(sample, "fn") %>% dplyr::mutate(fp_or_fn = "fn")

  dplyr::bind_rows(fp, fn) %>%
    dplyr::mutate(sample = sample, PCGR_TIER = as.factor(PCGR_TIER)) %>%
    dplyr::arrange(CHROM) %>%
    dplyr::select(sample, everything())
}

pcgr_results <-
  purrr::map(samples, read_pcgr_fpfn) %>%
  dplyr::bind_rows()
```

* Tier Summary:

```{r pcgr-tier-summary}
pcgr_results %>% pull(PCGR_TIER) %>% table() %>% addmargins() %>%
  tibble::as_tibble() %>%
  purrr::set_names(c("PCGR_TIER", "total")) %>%
  knitr::kable() %>%
  kableExtra::kable_styling(full_width = FALSE, position = "left")

pcgr_results %>% pull(sample) %>% table() %>% addmargins() %>%
  tibble::as_tibble() %>%
  purrr::set_names(c("samplename", "total fp+fn")) %>%
  knitr::kable() %>%
  kableExtra::kable_styling(full_width = FALSE, position = "left")
```

* FP/FN Table:

```{r pcgr-fpfn-table}
pcgr_results %>%
  DT::datatable(
    rownames = FALSE, caption = "PCGR FP and FN variants for all samples.",
    filter = list(position = "top", clear = FALSE, plain = TRUE),
    extensions = c("Scroller"),
    options = list(scroller = TRUE, scrollX = TRUE, scrollY = 220))
```



### Structural Variants

```{r funcs-svs}
read_manta_both <- function(sample, f1, f2) {

  read_manta_file <- function(sample, f1_or_f2) {
    stopifnot(sample %in% samples, f1_or_f2 %in% c(f1, f2))
    manta_fnames <- list.files(file.path(raijin_share, sample, f1_or_f2),
                               pattern = "prioritize.*vcf.gz$", recursive = TRUE, full.names = TRUE)
    manta_som_fname <- manta_fnames[!grepl("-germline-sv-prioritize", basename(manta_fnames))]
    stopifnot(length(manta_som_fname) == 1)

    sv <- manta_som_fname %>%
      normalizePath() %>%
      rock::prep_manta_vcf()

    list(sv = sv$sv, fname = manta_som_fname)
  }

  sv1 <- read_manta_file(sample, f1)
  sv2 <- read_manta_file(sample, f2)

  list(sv1 = list(sv = sv1$sv, fname = sv1$fname),
       sv2 = list(sv = sv2$sv, fname = sv2$fname))
}

manta_isec <- function(sample, f1, f2, bnd_switch = TRUE) {

  switch_bnd <- function(d) {
    stopifnot(all(colnames(d) %in% c("chrom1", "pos1", "chrom2", "pos2", "svtype")))
    no_bnd <- dplyr::filter(d, !.data$svtype %in% "BND")
    bnd <-
      dplyr::filter(d, .data$svtype %in% "BND") %>%
      dplyr::select(chrom1 = .data$chrom2, pos1 = .data$pos2,
                    chrom2 = .data$chrom1, pos2 = .data$pos1, .data$svtype)

    dplyr::bind_rows(no_bnd, bnd)
  }

  mb <- read_manta_both(sample, f1, f2)
  manta_fname1 <- mb$sv1$fname
  manta_fname2 <- mb$sv2$fname
  tot_vars <- list(run1 = nrow(mb$sv1$sv), run2 = nrow(mb$sv2$sv))
  col_nms <- colnames(mb$sv1$sv)
  fp <- dplyr::anti_join(mb$sv1$sv, mb$sv2$sv, by = col_nms)
  fn <- dplyr::anti_join(mb$sv2$sv, mb$sv1$sv, by = col_nms)

  if (bnd_switch) {
    # some real matching cases simply have switched BND mates
    fp_new <- dplyr::anti_join(switch_bnd(fp), fn, by = col_nms) %>% switch_bnd()
    fn_new <- dplyr::anti_join(switch_bnd(fn), fp, by = col_nms) %>% switch_bnd()
    fp <- fp_new
    fn <- fn_new
  }

  return(list(sample = sample,
              file1 = manta_fname1,
              file2 = manta_fname2,
              tot_vars = tot_vars,
              fp = fp,
              fn = fn))
}

manta_isec_stats <- function(mi) {
  samplename <- mi$sample
  fn <- nrow(mi$fn)
  fp <- nrow(mi$fp)
  tp <- mi$tot_vars$run2 - fn
  truth <- tp + fn
  recall <- tp / truth
  precision <- tp / (tp + fp)

  dplyr::tibble(
    sample = samplename, vcf_type = "manta-svpri",
    run1_count = mi$tot_vars$run1, run2_count = mi$tot_vars$run2,
    Recall = round(recall, 3), Precision = round(precision, 3), Truth = truth,
    TP = tp, FP = fp, FN = fn
  )
}

get_circos <- function(mi) {

  prep_svs_circos <- function(mi) {
    sv <- dplyr::bind_rows(list(fp = mi$fp, fn = mi$fn), .id = "fp_or_fn")
    links_coloured <- sv %>%
      dplyr::mutate(chrom1 = paste0("hs", .data$chrom1),
                    chrom2 = paste0("hs", .data$chrom2)) %>%
      dplyr::mutate(col = dplyr::case_when(
        fp_or_fn == "fp" ~ '(0,255,0)',
        fp_or_fn == "fn" ~ '(255,0,0)',
        TRUE ~ '0,0,0')) %>%
      dplyr::mutate(pos1b = .data$pos1,
                    pos2b = .data$pos2,
                    col = paste0('color=', col)) %>%
      dplyr::select(.data$chrom1, .data$pos1, .data$pos1b, .data$chrom2, .data$pos2, .data$pos2b, .data$col)
    return(links_coloured)
  }

  write_circos_configs <- function(mi) {
    sample <- mi$sample
    outdir <- file.path(raijin_share, sample, "woof/final/circos/manta_isec")
    dir.create(outdir, recursive = TRUE)

    links <- prep_svs_circos(mi)
    message(glue::glue("Writing circos links file to '{outdir}'."))
    readr::write_tsv(links, file.path(outdir, "SAMPLE.link.circos"), col_names = FALSE)

    message(glue::glue("Copying circos templates to '{outdir}'"))
    file.copy(from = system.file("templates/circos", glue::glue("circos_sv.conf"), package = "pebbles"),
              to = file.path(outdir, "circos.conf"), overwrite = TRUE)
    file.copy(system.file("templates/circos", "gaps.txt", package = "pebbles"), outdir, overwrite = TRUE)
    file.copy(system.file("templates/circos", "ideogram.conf", package = "pebbles"), outdir, overwrite = TRUE)

    return(outdir)
  }

  run_circos <- function(outdir, sample) {
    export_path <- "export PATH=/g/data3/gx8/extras/umccrise/miniconda/envs/umccrise_purple/bin:/g/data3/gx8/extras/umccrise/miniconda/bin:/opt/bin:/bin:/usr/bin:/opt/pbs/default/bin"
    cmd <- glue::glue("{export_path} && echo $(date) running circos on {sample} && circos -nosvg -conf {outdir}/circos.conf -outputdir {outdir} -outputfile circos_{sample}.png")
    system(cmd, ignore.stdout = TRUE)
    return(file.path(outdir, glue::glue("circos_{sample}.png")))
  }

  outdir <- write_circos_configs(mi)
  circos_png <- run_circos(outdir, mi$sample)
  circos_png
}


sv_results_per_sample <- function(sample) {
  mi <- manta_isec(sample, f1, f2)
  mi_stats <- manta_isec_stats(mi)
  circos <- get_circos(mi)

  list(mi = mi, stats = mi_stats, circos = circos)
}
```




```{r analysis-svs, message=F, warning=F}
sv_results <-
  purrr::map(samples, sv_results_per_sample) %>%
  purrr::set_names(samples)
```

```{r tab-descr-svs}
x <- dplyr::tribble(
  ~Column,    ~Description,
  "run1",     f1,
  "run2",     paste(f2, "(truthset)"),
  "FP",       "False Positive",
  "FN",       "False Negative",
  "TP",       "True Positive",
  "Truth",    "TP + FN",
  "Recall",   "TP / Truth",
  "Precision", "TP / TP + FP"
)

cap <- x %>%
  mutate(d = paste(Column, Description, sep = ": ")) %>%
  pull(d)
cap <- htmltools::tags$caption(htmltools::div(paste(cap[1:4], collapse = "; ")),
                               htmltools::div(paste(cap[5:length(cap)], collapse = "; ")))
```


#### Comparison Table

```{r eval-tab-svs}
bgsize <- '90% 90%'
big_mark_cols <- c("run1_count", "run2_count",
                   "Truth", "TP", "FP", "FN")

sv_stats <- purrr::map_dfr(sv_results, "stats")
sv_stats %>%
  dplyr::mutate_if(is.character, as.factor) %>%
  DT::datatable(
    rownames = FALSE, caption = cap,
    filter = list(position = "top", clear = FALSE, plain = TRUE),
    extensions = c("Scroller"),
    options = list(scroller = TRUE, scrollX = TRUE, scrollY = 220)) %>%
  formatStyle(
    'Recall',
    background = styleColorBar(sv_stats$Recall, 'lightgreen'),
    backgroundSize = bgsize,
    backgroundRepeat = 'no-repeat',
    backgroundPosition = 'center'
  ) %>%
  formatStyle(
    'Precision',
    background = styleColorBar(sv_stats$Precision, 'lightgreen'),
    backgroundSize = bgsize,
    backgroundRepeat = 'no-repeat',
    backgroundPosition = 'center'
  ) %>%
  formatCurrency(big_mark_cols, currency = "", interval = 3, mark = ",", digits = 0)
```

#### FP/FN Table
Showing FP and FN SVs from above:

```{r sv-fp-fn-table}
fp <- purrr::map(sv_results, function(x) x[["mi"]][["fp"]]) %>%
  dplyr::bind_rows(.id = "sample")
fn <- purrr::map(sv_results, function(x) x[["mi"]][["fn"]]) %>%
  dplyr::bind_rows(.id = "sample")
fp_fn <- dplyr::bind_rows(list(fp = fp, fn = fn), .id = "FP_or_FN")

fp_fn %>%
  dplyr::mutate_if(is.character, as.factor) %>%
  DT::datatable(
    rownames = FALSE,
    filter = list(position = "top", clear = FALSE, plain = TRUE),
    extensions = c("Scroller"),
    options = list(scroller = TRUE, scrollX = TRUE, scrollY = 400)) %>%
  formatCurrency(c("pos1", "pos2"), currency = "", interval = 3, mark = ",", digits = 0)
```


#### Circos Plots {.tabset .tabset-fade}

* <span style="color:green">Green</span> = False Positives
* <span style="color:red">Red</span> = False Negatives

```{r prep-sv-circos}
circos_plots <- purrr::map_chr(sv_results, "circos")
```

##### `r i <- 0; samples[i <- i + 1]`

```{r out.width="50%"}
knitr::include_graphics(circos_plots[i])
```

##### `r  samples[i <- i + 1]`

```{r out.width="50%"}
knitr::include_graphics(circos_plots[i])
```

##### `r  samples[i <- i + 1]`

```{r out.width="50%"}
knitr::include_graphics(circos_plots[i])
```

##### `r  samples[i <- i + 1]`

```{r out.width="50%"}
knitr::include_graphics(circos_plots[i])
```

##### `r  samples[i <- i + 1]`

```{r out.width="50%"}
knitr::include_graphics(circos_plots[i])
```


### Copy Number Variants

```{r purple-cnvs1}
read_purplegene_both <- function(sample, f1, f2) {

  read_purplegene_file <- function(sample, f1_or_f2) {
    stopifnot(sample %in% samples, f1_or_f2 %in% c(f1, f2))
    sample_sans_dots <- gsub("\\.", "_", sample)
    purplegene_fname <- list.files(normalizePath(Sys.glob(file.path(raijin_share, sample, f1_or_f2, "../umccrise*", paste0(sample_sans_dots, "*"), "purple"))),
                                   pattern = "purple.gene.cnv$", recursive = TRUE, full.names = TRUE)
    stopifnot(length(purplegene_fname) == 1)

    col_nms <- c(
      "Chromosome", "Start", "End", "Gene", "MinCopyNumber", "MaxCopyNumber",
      "SomaticRegions", "GermlineHomRegions", "GermlineHet2HomRegions",
      "TranscriptId", "TranscriptVersion", "ChromosomeBand",
      "MinRegions", "MinRegionStart", "MinRegionEnd", "MinRegionStartSupport",
      "MinRegionEndSupport", "MinRegionMethod",
      "NonsenseBiallelicCount", "NonsenseNonBiallelicCount",
      "NonsenseNonBiallelicPloidy", "SpliceBiallelicCount",
      "SpliceNonBiallelicCount", "SpliceNonBiallelicPloidy",
      "MissenseBiallelicCount", "MissenseNonBiallelicCount",
      "MissenseNonBiallelicPloidy", "MinMinorAllelePloidy", "ExonicBases")

    col_nms_reality <- readr::read_lines(purplegene_fname, n_max = 1) %>% strsplit("\\t") %>% unlist()
    stopifnot(all(col_nms == col_nms_reality[c(1:6, 8:length(col_nms_reality))]))

    cnv <- readr::read_tsv(purplegene_fname,
                           skip = 1,
                           col_names = col_nms,
                           col_types = cols(
                             .default = col_character(),
                             Chromosome = col_character(),
                             Start = col_double(),
                             End = col_double(),
                             Gene = col_character(),
                             MinCopyNumber = col_double(),
                             MaxCopyNumber = col_double())) %>%
      dplyr::select(Chromosome:MaxCopyNumber) %>%
      dplyr::mutate(MinCopyNumber = round(MinCopyNumber, 1),
                    MaxCopyNumber = round(MaxCopyNumber, 1))

    list(cnv = cnv, fname = purplegene_fname)
  }

  cnv1 <- read_purplegene_file(sample, f1)
  cnv2 <- read_purplegene_file(sample, f2)

  list(cnv1 = list(cnv = cnv1$cnv, fname = cnv1$fname),
       cnv2 = list(cnv = cnv2$cnv, fname = cnv2$fname))
}

cnv_results_per_sample <- function(sample, threshold = 0.5) {
  p_both <- read_purplegene_both(sample, f1, f2)


  # first check gene coordinates are identical
  pgene1 <- p_both$cnv1$cnv[c("Chromosome", "Start", "End", "Gene")]
  pgene2 <- p_both$cnv2$cnv[c("Chromosome", "Start", "End", "Gene")]
  stopifnot(identical(pgene1, pgene2))

  # Next check out min/max cn
  min_cn_diff <- tibble::tibble(
    chrom = p_both$cnv1$cnv$Chromosome,
    start = p_both$cnv1$cnv$Start,
    end = p_both$cnv1$cnv$End,
    gene = p_both$cnv1$cnv$Gene,
    cn1 = p_both$cnv1$cnv$MinCopyNumber,
    cn2 = p_both$cnv2$cnv$MinCopyNumber) %>%
    dplyr::mutate(equalish = abs(cn1 - cn2) < threshold,
                  samplename = sample) %>%
    dplyr::filter(!equalish)

  max_cn_diff <-
    tibble::tibble(
      chrom = p_both$cnv1$cnv$Chromosome,
      start = p_both$cnv1$cnv$Start,
      end = p_both$cnv1$cnv$End,
      gene = p_both$cnv1$cnv$Gene,
      cn1 = p_both$cnv1$cnv$MaxCopyNumber,
      cn2 = p_both$cnv2$cnv$MaxCopyNumber) %>%
    dplyr::mutate(equalish = abs(cn1 - cn2) < threshold,
                  samplename = sample) %>%
    dplyr::filter(!equalish)

  cn_diff <-
    dplyr::bind_rows(list(mincn = min_cn_diff, maxcn = max_cn_diff), .id = "min_or_max") %>%
    dplyr::select(samplename, min_or_max, chrom, start, end, gene, cn1, cn2, equalish)
  cn_diff
}

key_genes <- "/g/data3/gx8/extras/umccrise_2019_Feb/ngs-utils/ngs_utils/reference_data/key_genes/umccr_cancer_genes.2019-02-12.tsv" %>%
  readr::read_tsv(col_names = TRUE, col_types = cols(.default = col_character())) %>%
  dplyr::pull(symbol)

stopifnot(length(key_genes) == 1246)

cnv_results <-
  purrr::map(samples, cnv_results_per_sample) %>%
  dplyr::bind_rows() %>%
  dplyr::mutate(umccr_gene = gene %in% key_genes)

DT::datatable(cnv_results,
              caption = "Min/Max copynumber difference between the two runs",
              rownames = TRUE, filter = list(position = "top", clear = FALSE, plain = TRUE),
              extensions = c("Scroller"), options = list(scroller = TRUE, scrollX = TRUE, scrollY = 400))
```

Methods
-------
Here we're comparing the output from two bcbio runs:

- __Run1__: `r f1`
- __Run2__ (_"truthset"_): `r f2`


| Tumor/Normal Sample   |
|-----------------------|
| `2016.249.17.MH.P033` |
| `2016.249.18.SV.P007` |
| `2016.249.18.WH.P025` |
| `CUP-Pairs8`          |
| `SFRC01073`           |

* The comparison of VCFs has been done using the `compare` module of [woof](https://github.com/pdiakumis/woof).
* The `Rmd` template used to generate this report is available at <https://github.com/pdiakumis/compare-pipeline-runs>.

### SNVs/INDELs

- split into PASS and ALL variants
- run `bcftools isec` to generate VCFs containing variants that are:
  - False Positives (FPs): within Run1, but not Truthset
  - False Negatives (FNs): within Truthset, but not Run1
  - True Positives (TPs): within _both_ Run1 and Truthset
- run `eval_vcf` function in <https://github.com/umccr/vcf_stuff/>
  to generate a summary table with evaluation statistics

#### Diagram

```{r graph1, fig.height=7}
set.seed(42)

ndf <-
  tribble(
    ~id, ~label, ~type,
    1, "snv1", "file-initial",
    2, "snv2", "file-initial",
    3, "bcftools-view-f", "command",
    4, "vcf1_pass", "file",
    5, "vcf2_pass", "file",
    6, "bcftools-isec", "command",
    7, "FP-FN-TP", "file",
    8, "FP-FN-TP_pass", "file",
    9, "count-vars", "command",
    10, "counts", "file",
    11, "counts_pass", "file",
    12, "eval-vcf", "command") %>%
  mutate(
    shape = case_when(
      type == "file" ~ "rectangle",
      type == "command" ~ "circle",
      TRUE ~ "square"
    ),
    fillcolor = case_when(
      type == "file" ~ "lightblue",
      type == "command" ~ "#f48f42",
      TRUE ~ "#eef442"
    ),
    fontcolor = "black")

edf <-
  create_edge_df(
    from = c(1, 2, 3, 3, 1, 2, 4, 5, 6, 6, 1, 2, 4, 5, 9, 9, 7, 8, 10, 11),
    to =   c(3, 3, 4, 5, 6, 6, 6, 6, 7, 8, 9, 9, 9, 9, 10, 11, 12, 12, 12, 12)
  )


create_graph(nodes_df = ndf, edges_df = edf) %>%
  set_node_attrs(
    node_attr = "fontsize",
    values = "8"
  ) %>%
  render_graph()
```

### Structural Variants

- keep all variants from Run1 and Run2 (i.e. don't filter any variants)
- generate
  - number of SVs in Run1
  - number of SVs in Run2
  - FP variants: within Run1, but not Truthset
  - FN variants: within Truthset, but not Run1
- generate one circos plot per sample, where:
  - FN in red
  - FP in green

### Copy Number Variants

- check if gene number and coordinates are identical between runs
- check which genes have a difference in Min/Max copy number greater than 0.5
