---
author: "University of Melbourne Centre for Cancer Research"
date: "`r format(Sys.time(), '%a %Y-%b-%d %H:%M')`"
output:
  html_document:
    theme: simplex
    toc: true
    toc_float: false
    toc_depth: 5
    code_download: true
    code_folding: hide
  rmdformats::material:
    highlight: kate
params:
  woof_final: "/path/to/woof/final"
  run1_nm: "run1"
  run2_nm: "run2"
title: "`r glue::glue('Comparison of woof runs in {htmltools::br()}\'{normalizePath(params$woof_final)}\':{htmltools::br()}{params$run1_nm}{htmltools::br()}  vs.{htmltools::br()}{params$run2_nm}')`"
description: "Comparison of bcbio and umccrise pipeline runs"
---

<style type="text/css">
.main-container {
  max-width: 2000px;
  margin-left: auto;
  margin-right: auto;
}
</style>


```{r load_knitr, include=F}
require(knitr)
knitr::opts_chunk$set(
  echo = FALSE
  )
```

```{r render_report_interactively, eval=F, echo=F}
params_tmp <- list(
  local = list(
    woof_final = "/Users/pdiakumis/Desktop/projects/umccr/woof_compare/test/woof/final",
    run1_nm = "run1",
    run2_nm = "run2"
  ),
  gadi = list(
    woof_final = "",
    run1_nm = "run1",
    run2_nm = "run2"
  )
)
params <- params_tmp[["local"]]

rmd <- here::here("inst/rmd/woof_compare.Rmd")
render_me <- function() {
  rmarkdown::render(
    rmd,
    params = params)
}
render_me()
```

```{r load_pkgs, message=F, warning=F}
suppressPackageStartupMessages(library(bedr))
suppressPackageStartupMessages(library(DiagrammeR))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(DT))
suppressPackageStartupMessages(library(glue))
suppressPackageStartupMessages(library(kableExtra))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(readr))
suppressPackageStartupMessages(library(rock))
suppressPackageStartupMessages(library(tidyr))
suppressPackageStartupMessages(library(woofr))
```

Results
-------

### SNVs/Indels

```{r funcs_snvs}
woof_final <- params$woof_final
stopifnot(dir.exists(woof_final))

samples <- list.dirs(woof_final, recursive = FALSE, full.names = FALSE)
# get vcf_types from unique woof/final/<sample>/vcf_eval/<vcf_type>/
snv_vcf_types <-
  file.path(woof_final, samples, "vcf_eval") %>%
  list.dirs(recursive = FALSE, full.names = FALSE) %>%
  unique()
allowed_snv_vcf_types <- c("ensemble-batch", "ensemble-germ", "gatk-germ", "mutect-batch",
                           "strelka-batch", "strelka-germ", "vardict-batch", "vardict-germ",
                           "pcgr", "cpsr")
stopifnot(snv_vcf_types %in% allowed_snv_vcf_types)

read_count_files <- function(woof_final, sample, vcf_type, pass_or_all) {

  c1 <- file.path(woof_final, sample, "vcf_counts", "f1", vcf_type, pass_or_all, "count_vars.txt") %>% woofr::read_count_file()
  c2 <- file.path(woof_final, sample, "vcf_counts", "f2", vcf_type, pass_or_all, "count_vars.txt") %>% woofr::read_count_file()

  # keeping sample/vcf_type/subset in row
  dplyr::tibble(sample = sample, vcf_type = vcf_type, subset = pass_or_all, run1_count = c1, run2_count = c2)
}

get_stats <- function(woof_final, sample, vcf_type, pass_or_all) {
  dplyr::bind_cols(
    read_count_files(woof_final, sample, vcf_type, pass_or_all),
    woofr::read_eval_file(file.path(woof_final, sample, "vcf_eval", vcf_type, pass_or_all, "eval_stats.tsv")))
}

get_stats_sample <- function(sample, vcf_types) {

  get_stats_sample_subset <- function(sample, pass_or_all) {
    vcf_types %>%
      purrr::map(~ get_stats(woof_final, sample, .x, pass_or_all)) %>%
      dplyr::bind_rows()
  }

  dplyr::bind_rows(
    get_stats_sample_subset(sample, "PASS"),
    get_stats_sample_subset(sample, "ALL")
  )
}
```

```{r analysis_snvs}
d <- purrr::map(samples, ~ get_stats_sample(.x, snv_vcf_types)) %>%
  dplyr::bind_rows() %>%
  dplyr::mutate(id = dplyr::row_number())
```

```{r tab_descr_snvs}
x <- dplyr::tribble(
  ~Column,    ~Description,
  "run1",     "woof run1",
  "run2",     "woof run2 (truthset)",
  "subset",   "All/PASSed variants",
  "SNP",      "SNP results",
  "IND",      "Indel results",
  "FP",       "False Positive",
  "FN",       "False Negative",
  "TP",       "True Positive",
  "Truth",    "TP + FN",
  "Recall",   "TP / Truth",
  "Precision", "TP / TP + FP"
)

cap <- x %>%
  mutate(y = paste(Column, Description, sep = ": ")) %>%
  pull(y)
cap <- htmltools::tags$caption(htmltools::div(paste(cap[1:5], collapse = "; ")),
                               htmltools::div(paste(cap[6:length(cap)], collapse = "; ")))
```

#### Comparison Table

```{r eval_tab_snvs}
bgsize <- '90% 90%'
big_mark_cols <- c("run1_count", "run2_count",
                   "SNP_Truth", "SNP_TP", "SNP_FP", "SNP_FN",
                   "IND_Truth", "IND_TP", "IND_FP", "IND_FN")
bar_colour <- "lightblue"
bar_range <- c(0.97, 1)

d %>%
  dplyr::select(-matches("f[123]$")) %>%
  dplyr::select(id, sample, vcf_type, subset, run1_count, run2_count,
                contains("Recall"), contains("Precision"),
                everything()) %>%
  dplyr::mutate_if(is.numeric, round, 3) %>%
  dplyr::mutate_if(is.character, as.factor) %>%
  DT::datatable(
    rownames = FALSE, caption = cap, class = "cell-border stripe hover compact",
    filter = list(position = "top", clear = FALSE, plain = FALSE),
    extensions = c("Scroller", "KeyTable"),
    options = list(scroller = TRUE, scrollY = 600, autoWidth = TRUE, keys = TRUE)
  ) %>%
  formatStyle(
    'SNP_Recall',
    background = styleColorBar(bar_range, color = bar_colour),
    backgroundSize = bgsize, backgroundRepeat = 'no-repeat', backgroundPosition = 'center'
  ) %>%
  formatStyle(
    'IND_Recall',
    background = styleColorBar(bar_range, color = bar_colour),
    backgroundSize = bgsize, backgroundRepeat = 'no-repeat', backgroundPosition = 'center'
  ) %>%
  formatStyle(
    'SNP_Precision',
    background = styleColorBar(bar_range, color = bar_colour),
    backgroundSize = bgsize, backgroundRepeat = 'no-repeat', backgroundPosition = 'center'
  ) %>%
  formatStyle(
    'IND_Precision',
    background = styleColorBar(bar_range, color = bar_colour),
    backgroundSize = bgsize, backgroundRepeat = 'no-repeat', backgroundPosition = 'center'
  ) %>%
  formatCurrency(big_mark_cols, currency = "", interval = 3, mark = ",", digits = 0)
```

Methods
-------
Here we're comparing the output from two bcbio runs:

- __Run1__: `r params$run1_nm`
- __Run2__ (_"truthset"_): `r params$run2_nm`

* The comparison of VCFs has been done using the `compare` module of [woof](https://github.com/umccr/woof).
* The `Rmd` template used to generate this report is available in [woofr](https://github.com/umccr/woofr).

### SNVs/INDELs

- split into PASS and ALL variants
- run `bcftools isec` to generate VCFs containing variants that are:
  - False Positives (FPs): within Run1, but not Truthset
  - False Negatives (FNs): within Truthset, but not Run1
- True Positives (TPs): within _both_ Run1 and Truthset
- run `eval_vcf` function in <https://github.com/umccr/vcf_stuff/>
  to generate a summary table with evaluation statistics

#### Diagram

```{r graph1, fig.height=7}
set.seed(42)

ndf <-
  tribble(
    ~id, ~label, ~type,
    1, "snv1", "file-initial",
    2, "snv2", "file-initial",
    3, "bcftools-view-f", "command",
    4, "vcf1_pass", "file",
    5, "vcf2_pass", "file",
    6, "bcftools-isec", "command",
    7, "FP-FN-TP", "file",
    8, "FP-FN-TP_pass", "file",
    9, "count-vars", "command",
    10, "counts", "file",
    11, "counts_pass", "file",
    12, "eval-vcf", "command") %>%
  mutate(
    shape = case_when(
      type == "file" ~ "rectangle",
      type == "command" ~ "circle",
      TRUE ~ "square"
    ),
    fillcolor = case_when(
      type == "file" ~ "lightblue",
      type == "command" ~ "#f48f42",
      TRUE ~ "#eef442"
    ),
    fontcolor = "black")

edf <-
  create_edge_df(
    from = c(1, 2, 3, 3, 1, 2, 4, 5, 6, 6, 1, 2, 4, 5, 9, 9, 7, 8, 10, 11),
    to =   c(3, 3, 4, 5, 6, 6, 6, 6, 7, 8, 9, 9, 9, 9, 10, 11, 12, 12, 12, 12)
  )


create_graph(nodes_df = ndf, edges_df = edf) %>%
  set_node_attrs(
    node_attr = "fontsize",
    values = "8"
  ) %>%
  render_graph()
```

### Structural Variants

- keep all variants from Run1 and Run2 (i.e. don't filter any variants)
- generate
  - number of SVs in Run1
  - number of SVs in Run2
  - FP variants: within Run1, but not Truthset
  - FN variants: within Truthset, but not Run1
- generate one circos plot per sample, where:
  - FN in red
  - FP in green

### Copy Number Variants

- check if gene number and coordinates are identical between runs
- check which genes have a difference in Min/Max copy number greater than 0.5
